Sources/SwiftBuildSupport/PIFBuilder.swift

```diff
@@ -144,7 +258,133 @@ public final class PIFBuilder {
+                            accessibleTools: accessibleTools,
+                            writableDirectories: writableDirectories,
+                            readOnlyDirectories: readOnlyDirectories,
+                            allowNetworkConnections: [],
+                            pkgConfigDirectories: self.pkgConfigDirectories,
+                            sdkRootPath: buildParameters.toolchain.sdkRootPath,
+                            fileSystem: fileSystem,
+                            modulesGraph: self.graph,
+                            observabilityScope: observabilityScope
+                        )
+
+                        let diagnosticsEmitter = observabilityScope.makeDiagnosticsEmitter {
+                            var metadata = ObservabilityMetadata()
+                            metadata.moduleName = module.name
+                            metadata.pluginName = result.plugin.name
+                            return metadata
+                        }
+
+                        for line in result.textOutput.split(whereSeparator: { $0.isNewline }) {
+                            diagnosticsEmitter.emit(info: line)
+                        }
+
+                        for diag in result.diagnostics {
+                            diagnosticsEmitter.emit(diag)
+                        }
+
+                        let result2 = PackagePIFBuilder.BuildToolPluginInvocationResult(
+                            prebuildCommandOutputPaths: result.prebuildCommands.map( { $0.outputFilesDirectory } ),
+                            buildCommands: result.buildCommands.map( { buildCommand in
+                                var env: [String: String] = [:]
+                                for (key, value) in buildCommand.configuration.environment {
+                                    env[key.rawValue] = value
+                                }
+
+                                return PackagePIFBuilder.CustomBuildCommand(
+                                    displayName: buildCommand.configuration.displayName,
+                                    executable: buildCommand.configuration.executable.pathString,
+                                    arguments: buildCommand.configuration.arguments,
+                                    environment: env,
+                                    workingDir: buildCommand.configuration.workingDirectory,
+                                    inputPaths: buildCommand.inputFiles,
+                                    outputPaths: buildCommand.outputFiles.map(\.pathString),
+                                    sandboxProfile: nil
+                                )
+                            } )
+                        )
+
+                        // Add a BuildToolPluginInvocationResult to the mapping.
+                        buildToolPluginResults.append(result2)
+                        buildtoolPluginResults[module.name] = result2
```

Comment:
Nit: the close naming here seems tricky to parse 😬.

Maybe rename `buildtoolPluginResults`  to `buildToolPluginResultsByModuleName`?